neqo-transport/src/connection/mod.rs

```diff
@@ -3074,14 +3078,23 @@ impl Connection {
         }
     }
 
-    fn decode_ack_delay(&self, v: u64) -> Duration {
+    fn decode_ack_delay(&self, v: u64) -> Res<Duration> {
         // If we have remote transport parameters, use them.
         // Otherwise, ack delay should be zero (because it's the handshake).
         self.tps.borrow().remote.as_ref().map_or_else(
-            || Duration::new(0, 0),
+            || Ok(Duration::default()),
             |r| {
-                let exponent = u32::try_from(r.get_integer(tparams::ACK_DELAY_EXPONENT)).unwrap();
-                Duration::from_micros(v.checked_shl(exponent).unwrap_or(u64::MAX))
+                let exponent = u32::try_from(r.get_integer(tparams::ACK_DELAY_EXPONENT))?;
+                if exponent > 20 {
```

Comment:
Might need a test for this.  Or, given that we have already checked that this is not > 20, an unwrap is OK.